cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -0,0 +1,604 @@
+	}
+
+	machineResource, _ := schema.ParseResourceArg(fmt.Sprintf("machines.%v", defaultMachineAPI))
+	if machineResource == nil {
+		panic("machineResource")
+	}
+	machineInformer := informerFactory.ForResource(*machineResource)
+	machineSetInformer := informerFactory.ForResource(*machineSetResource)
+	var machineDeploymentInformer informers.GenericInformer
+
+	machineDeploymentInformer = informerFactory.ForResource(*machineDeploymentResource)
+	machineDeploymentInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{})
+
+	machineInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{})
+	machineSetInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{})
+
+	nodeInformer := kubeInformerFactory.Core().V1().Nodes().Informer()
+	nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{})
+
+	if err := machineInformer.Informer().GetIndexer().AddIndexers(cache.Indexers{
+		machineProviderIDIndex: indexMachineByProviderID,
+	}); err != nil {
+		return nil, fmt.Errorf("cannot add machine indexer: %v", err)
+	}
+
+	if err := nodeInformer.GetIndexer().AddIndexers(cache.Indexers{
+		nodeProviderIDIndex: indexNodeByProviderID,
+	}); err != nil {
+		return nil, fmt.Errorf("cannot add node indexer: %v", err)
+	}
+
+	return &machineController{
+		kubeInformerFactory:       kubeInformerFactory,
+		machineInformerFactory:    informerFactory,
+		machineDeploymentInformer: machineDeploymentInformer,
+		machineInformer:           machineInformer,
+		machineSetInformer:        machineSetInformer,
+		nodeInformer:              nodeInformer,
+		dynamicclient:             dynamicclient,
+		machineSetResource:        machineSetResource,
+		machineResource:           machineResource,
+		machineDeploymentResource: machineDeploymentResource,
+	}, nil
+}
+
+func (c *machineController) machineSetProviderIDs(machineSet *MachineSet) ([]string, error) {
+	machines, err := c.machinesInMachineSet(machineSet)
+	if err != nil {
+		return nil, fmt.Errorf("error listing machines: %v", err)
+	}
+
+	var providerIDs []string
+	for _, machine := range machines {
+		if machine.Spec.ProviderID == nil || *machine.Spec.ProviderID == "" {
+			klog.Warningf("Machine %q has no providerID", machine.Name)
+		}
+
+		if machine.Spec.ProviderID != nil && *machine.Spec.ProviderID != "" {
+			providerIDs = append(providerIDs, *machine.Spec.ProviderID)
+			continue
+		}
+
+		if machine.Status.NodeRef == nil {
+			klog.V(4).Infof("Status.NodeRef of machine %q is currently nil", machine.Name)
+			continue
+		}
+
+		if machine.Status.NodeRef.Kind != "Node" {
+			klog.Errorf("Status.NodeRef of machine %q does not reference a node (rather %q)", machine.Name, machine.Status.NodeRef.Kind)
+			continue
+		}
+
+		node, err := c.findNodeByNodeName(machine.Status.NodeRef.Name)
+		if err != nil {
+			return nil, fmt.Errorf("unknown node %q", machine.Status.NodeRef.Name)
+		}
+
+		if node != nil {
+			providerIDs = append(providerIDs, node.Spec.ProviderID)
```

Comment:
This seems like it will duplicate providerIDs in most cases here, since they *should* match between the Machine and the Node (though one may be more specific than the other, if the providerID for the node also includes information like region, etc).

Reply:
Let me look through `cluster-autoscaler/cloudprovider/*` to see if there's an expectation here.

Reply:
A few lines above if we match we continue: 

```go
		if machine.Spec.ProviderID != nil && *machine.Spec.ProviderID != "" {
			providerIDs = append(providerIDs, *machine.Spec.ProviderID)
			continue
		}
```
Does that answer the duplication? Maybe I'm missing something more subtle in your observation.

Reply:
> This seems like it will duplicate providerIDs in most cases here, since they _should_ match between the Machine and the Node (though one may be more specific than the other, if the providerID for the node also includes information like region, etc).

I addressed this in commit https://github.com/kubernetes/autoscaler/pull/1866/commits/c21f987b1403d88087a84639d46d1210cda0529b. That normalizes the providerID values.

Reply:
Thanks for clarifying